internal/engine/selectors/selectors.go

```diff
@@ -0,0 +1,372 @@
+
+	return &EntitySelection{
+		selector: selector,
+		entity:   entityType,
+	}, nil
+}
+
+// compileSelectorForEntity compiles a selector expression for a given entity type into a CEL program
+func (e *Env) compileSelectorForEntity(selector string, entityType minderv1.Entity) (compiledSelector, error) {
+	env, err := e.envForEntity(entityType)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get environment for entity %v: %w", entityType, err)
+	}
+
+	ast, issues := env.Parse(selector)
+	if issues.Err() != nil {
+		return nil, fmt.Errorf("failed to parse expression %q: %w", selector, issues.Err())
+	}
+
+	checked, issues := env.Check(ast)
+	if issues.Err() != nil {
+		return nil, fmt.Errorf("failed to check expression %q: %w", selector, issues.Err())
+	}
+
+	program, err := env.Program(checked)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create program for expression %q: %w", selector, err)
+	}
+
+	return program, nil
+}
+
+// envForEntity gets the CEL environment for a given entity type. If the environment is not cached,
+// it creates it using the factory for that entity type.
+func (e *Env) envForEntity(entity minderv1.Entity) (*cel.Env, error) {
+	cache, ok := e.entityEnvs[entity]
+	if !ok {
+		return nil, fmt.Errorf("no cache found for entity %v", entity)
+	}
+
+	cache.once.Do(func() {
+		cache.env, cache.err = e.factories[entity].createEnv()
+	})
+
+	return cache.env, cache.err
+}
+
+// entityInfoConverter is an interface for converting an entity from an EntityInfoWrapper to a SelectorEntity
+type entityInfoConverter interface {
+	toSelectorEntity(entity proto.Message) *minderv1.SelectorEntity
+}
+
+type repositoryInfoConverter struct{}
+
+func (_ *repositoryInfoConverter) toSelectorEntity(entity proto.Message) *minderv1.SelectorEntity {
+	r, ok := entity.(*minderv1.Repository)
+	if !ok {
+		return nil
+	}
+
+	return &minderv1.SelectorEntity{
+		EntityType: minderv1.Entity_ENTITY_REPOSITORIES,
+		Name:       fmt.Sprintf("%s/%s", r.GetOwner(), r.GetName()),
+		Entity: &minderv1.SelectorEntity_Repository{
+			Repository: &minderv1.SelectorEntity_SelectorRepository{
+				Name:      fmt.Sprintf("%s/%s", r.GetOwner(), r.GetName()),
+				RepoName:  r.GetName(),
+				Owner:     r.GetOwner(),
+				IsFork:    proto.Bool(r.GetIsFork()),
+				IsPrivate: proto.Bool(r.GetIsPrivate()),
+			},
+		},
+	}
+}
+
+type artifactInfoConverter struct{}
+
+func (_ *artifactInfoConverter) toSelectorEntity(entity proto.Message) *minderv1.SelectorEntity {
+	a, ok := entity.(*minderv1.Artifact)
+	if !ok {
+		return nil
+	}
+
+	return &minderv1.SelectorEntity{
+		EntityType: minderv1.Entity_ENTITY_ARTIFACTS,
+		Name:       fmt.Sprintf("%s/%s", a.GetOwner(), a.GetName()),
+		Entity: &minderv1.SelectorEntity_Artifact{
+			Artifact: &minderv1.SelectorEntity_SelectorArtifact{
+				Name:         fmt.Sprintf("%s/%s", a.GetOwner(), a.GetName()),
+				ArtifactName: a.GetName(),
+				Owner:        a.GetOwner(),
+				Type:         a.GetType(),
+			},
+		},
+	}
```

Comment:
**question:** it feels to me like this piece of code belongs to `EntityInfoWrapper` rather than to `Selector`, but I am not 100% sure.

I suppose that every entity managed by Minder engine will eventually need this sort of transformation, and adding a new entity would lead up to changes to Selector. If my reasoning holds, it is the `Entity` being aware of `Selector`, and not the other way around, as `Selector` is really not much more than a boolean predicate.

What's your thought?

Reply:
> **question:** it feels to me like this piece of code belongs to `EntityInfoWrapper` rather than to `Selector`, but I am not 100% sure.

I was wondering about this, too when working on the code. I don't have a good answer (as you can see below)

> 
> I suppose that every entity managed by Minder engine will eventually need this sort of transformation, and adding a new entity would lead up to changes to Selector. If my reasoning holds, it is the `Entity` being aware of `Selector`, and not the other way around, as `Selector` is really not much more than a boolean predicate.
> 

The selector does expose some entity-specific attributes now like `is_fork`, so there is some coupling between the entity and the selector now and we'd still need to change the selector protobuf message if we add a new entity.

The protobuf messages though are just a convenience in general we could use a `map[string]any` (and that's what we pass to the evaluator), it's just that the protobuf messages make it easier to check the expressions and the input. 

Would you say that having the EntityInfoWrapper code be able to return the `SelectorEntity` (or actually..even returning the `map[string]any`...) to be treated as an opaque blob by the selector module would be better?

The other question I was wondering about - we will have to call the provider to fill in the key-value attributes (which are not present in the protobuf message now) and format the name.

Is this better done from the selector code that knows the expressions? Maybe this is premature optimization, but I wondered if we could only call the provider methods lazily if the attributes are needed in the expression.

Reply:
Hey @jhrozek, sorry for the late reply!

> Would you say that having the EntityInfoWrapper code be able to return the `SelectorEntity` (or actually..even returning the `map[string]any`...) to be treated as an opaque blob by the selector module would be better?

I think this is exactly the case. My reasoning is that selectors are expressed in a language that is specific to Minder's domain, while entities are Minder's view into someone else's domain. Entities should specify themselves "in Minder's lingo", and not the other way around.

This is also reflected by my previous comment about adding a new entity. Adding support for a new entity should only require minimal (or no) changes to the engine.

> The other question I was wondering about - we will have to call the provider to fill in the key-value attributes (which are not present in the protobuf message now) and format the name.
>
> Is this better done from the selector code that knows the expressions?

I'm not sure about this, but I would gather all the info up front and avoid getting information lazily for now.

Reply:
It feels like the selector should be using the [Visitor Pattern](https://en.wikipedia.org/wiki/Visitor_pattern) at a naive first pass, but I could be convinced otherwise.

That feels like it would like exposing a few methods that take attribute paths, ala: `StringProperty(string path) *string`, `BoolProperty(string path) *bool`, `IntProperty(string path) *int`.  (I'm not sure whether we should have both `IntProperty` and `FloatProperty` -- it feels like most properties right now are integer, but something like a Trusty score would be a float (I'm not suggesting we make this a property of Entities, but trying to think of examples).

Reply:
Yes, getting the key-value items from the provider was the plan all along, but you're right that we don't actually have to keep the conversion in the selector code.

This one though:
```
// Either the provider or some common code provides short-names for names like 'github.is_fork' --> 'is_fork'.
```
But we won't allow shortnames for *all* the key-values, but only for those that have an explicit attribute in the message right?

In other words, for a  a repository we might be able to use the following in the grammar: `repository.is_fork = false` which happens to be set to the same value as `repository.provider_properties.github.is_fork` but we can't use `repository.maturity=stable` even if our provider sets `repository.provider_properties.github.maturity=stable`?

Reply:
I'm not sure how we want to handle the shortnames, to be honest.  I can see two paths:

1. The shortnames are a convention, and it's up to providers (possibly with library help) to maintain that convention.
2. The shortnames are a minder intrinsic (e.g. produce a shortname for each fullname that has a unique last segment).  In this case, I'm imagining that `repository.is_fork` could be `is_fork` unless you also have a `github.tags.is_fork` (i.e. an "is fork" repo tag in the organization).

Looking at the second example, it feels like maybe we should go with case 1, which would also make it easier to report on the general set of properties which might be present, though we'd some syntax for `github.tags.* => string` to indicate that the GitHub tags were strings.

Reply:
In this version, I moved the logic that constructs the `SelectorEntity` to providers. I don't think we're done yet and I think we want the providers to also grab the key-value pairs and keep them as a map, but I'd prefer to do that in a follow-up PR if that's OK.

The attributes that are in the messages now are "stable" as far as the messages go, meaning I'd like to document them for users as something they can always use in the selector grammar. In the follow-up, I'd like to work on reading provider-specific key-value pairs per provider. First, I'd like to support languages for repos and the key-value pairs that github support (can't remember the names, but it's really a generic k-v pairs). At the same time, I'd like to work on "duck-typing" the attributes so that users can type `repository.properties.foo=bar` instead of `repository.provider.github.foo=bar`. But I think that merits a separate discussion.

Reply:
This version adds properties to repositories. There's no fetching of them in the executor but I think we don't need to show how to update a map of string,string.